toolchain/check/impl_validation.cpp

```diff
@@ -0,0 +1,456 @@
+  }
+
+  llvm::sort(interfaces_to_import);
+  llvm::unique(interfaces_to_import);
+
+  struct ImplToImport {
+    SemIR::ImportIRId ir_id;
+    SemIR::ImplId import_impl_id;
+
+    constexpr auto operator==(const ImplToImport& rhs) const -> bool = default;
+    constexpr auto operator<=>(const ImplToImport& rhs) const -> auto {
+      if (ir_id != rhs.ir_id) {
+        return ir_id.index <=> rhs.ir_id.index;
+      }
+      return import_impl_id.index <=> rhs.import_impl_id.index;
+    }
+  };
+
+  llvm::SmallVector<ImplToImport> impls_to_import;
+  for (auto [ir_id, interface_id] : interfaces_to_import) {
+    const SemIR::File& sem_ir = *context.import_irs().Get(ir_id).sem_ir;
+    for (auto [impl_id, impl] : sem_ir.impls().enumerate()) {
+      if (impl.is_final && impl.interface.interface_id == interface_id) {
+        impls_to_import.push_back({.ir_id = ir_id, .import_impl_id = impl_id});
+      }
+    }
+  }
+
+  llvm::sort(impls_to_import);
+  llvm::unique(impls_to_import);
+
+  for (auto [ir_id, import_impl_id] : impls_to_import) {
+    ImportImpl(context, ir_id, import_impl_id);
+  }
+}
+
+auto ValidateImplsInFile(Context& context) -> void {
+  ImportFinalImplsWithImplInFile(context);
+
+  // Collect all of the impls sorted into contiguous segments by their
+  // interface. We only need to compare impls within each such segment. We don't
+  // keep impls with an Error in them, as they may be missing other values
+  // needed to check the diagnostics and they already have a diagnostic printed
+  // about them anyhow. We also verify the impl has an `InterfaceId` since it
+  // can be missing, in which case a diagnostic would have been generated
+  // already as well.
+  //
+  // Don't hold Impl pointers here because the process of looking for
+  // diagnostics may cause imports and may invalidate pointers into the
+  // ImplStore.
+  llvm::SmallVector<ImplAndInterface> impl_ids_by_interface(llvm::map_range(
+      llvm::make_filter_range(
+          context.impls().enumerate(),
+          [](std::pair<SemIR::ImplId, const SemIR::Impl&> pair) {
+            return pair.second.witness_id != SemIR::ErrorInst::InstId &&
+                   pair.second.interface.interface_id.has_value();
+          }),
+      [](std::pair<SemIR::ImplId, const SemIR::Impl&> pair) {
+        return ImplAndInterface{.impl_id = pair.first,
+                                .interface = pair.second.interface};
+      }));
+  llvm::stable_sort(impl_ids_by_interface, [](const ImplAndInterface& lhs,
+                                              const ImplAndInterface& rhs) {
+    return lhs.interface.interface_id.index < rhs.interface.interface_id.index;
+  });
+
+  const auto* it = impl_ids_by_interface.begin();
+  while (it != impl_ids_by_interface.end()) {
+    const auto* segment_begin = it;
+    do {
+      ++it;
+    } while (it != impl_ids_by_interface.end() &&
+             it->interface.interface_id ==
+                 segment_begin->interface.interface_id);
```

Comment:
Don't know if it is worth trying to avoid indirection here, but it ends up being the same number of lines.
```suggestion
    const auto* segment_begin = it;
    auto interface_id = it->interface.interface_id;
    do {
      ++it;
    } while (it != impl_ids_by_interface.end() &&
             it->interface.interface_id == interface_id);
```

Reply:
Done